### PR TITLE
feature/nats

### PR DESCRIPTION
**Pull request by DmitryOnopriienko**
@ajax-sushkevych-s

### CODE REVIEW

File: wm-system/src/test/kotlin/com/ajaxproject/warehouse/controller/nats/product/CreateProductNatsControllerTest.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    @Test
+    fun testCreatesProductFromValidRequest() {
+        // GIVEN
+        val createProductRequest = CreateProductRequest.newBuilder()
+            .setTitle("Test create product")
+            .setPrice(19.99)
+            .setAmount(200)
+            .setAbout("New test product")
+            .build()
```
Comment (ajax-sushkevych-s) [inline wm-system/src/test/kotlin/com/ajaxproject/warehouse/controller/nats/product/CreateProductNatsControllerTest.kt]: Please refactor whenever you use the builder pattern in all cases
```suggestion
val createProductRequest = CreateProductRequest.newBuilder().apply {
            title = "Test create product"
            price = 19.99
            amount = 200
            about = "New test product"
        }.build()
```

File: wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt [hunk 2]
```diff
@@ -55,7 +55,7 @@ class WaybillServiceImpl(
             .map { it.productId as String }
             .minus(validIdList.toSet())
             .map { "Product with id $it not found" }
-            .toList()
+            .toList()   // TODO add product amount in DB change
```
Comment (ajax-sushkevych-s) [inline wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt]: Please, do not commit comments.
```suggestion
            .toList()
```

Comment (DmitryOnopriienko) [reply wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt]: Serhii said that its OK to commit TODOs

File: wm-system/src/test/kotlin/com/ajaxproject/warehouse/controller/nats/product/CreateProductNatsControllerTest.kt (hunk 1 above)
Comment (DmitryOnopriienko) [reply wm-system/src/test/kotlin/com/ajaxproject/warehouse/controller/nats/product/CreateProductNatsControllerTest.kt]: done

File: internal-api/src/main/proto/com/ajaxproject/api/internal/warehousesvc/input/reqreply/product/delete.proto [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    Success success = 1;
+    Failure failure = 2;
+  }
+
+  message Success {}
+
+  message Failure {
+    string message = 100;
+  }
+}
```
Comment (ajax-sushkevych-s) [inline internal-api/src/main/proto/com/ajaxproject/api/internal/warehousesvc/input/reqreply/product/delete.proto]: add empty line after that

Comment (DmitryOnopriienko) [reply internal-api/src/main/proto/com/ajaxproject/api/internal/warehousesvc/input/reqreply/product/delete.proto]: done

File: wm-system/src/main/kotlin/com/ajaxproject/warehouse/dto/ProductDataDto.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    val amount: Int,
+    val about: String?
+) {
+    fun mapToProto(): Product = Product.newBuilder()
+        .setId(id)
+        .setTitle(title)
+        .setPrice(price)
+        .setAmount(amount)
+        .setAbout(about)
+        .build()
```
Comment (ajax-sushkevych-s) [inline wm-system/src/main/kotlin/com/ajaxproject/warehouse/dto/ProductDataDto.kt]: Just a suggestion
```suggestion
    fun mapToProto(): Product = Product.newBuilder().apply {
        id = this@ProductDataDto.id
        title = this@ProductDataDto.title
        price = this@ProductDataDto.price
        amount = this@ProductDataDto.amount
        about = this@ProductDataDto.about
    }.build()
```

Comment (DmitryOnopriienko) [reply wm-system/src/main/kotlin/com/ajaxproject/warehouse/dto/ProductDataDto.kt]: It is so Kotlin-like, that it's not even beautiful :)

File: wm-system/src/main/kotlin/com/ajaxproject/warehouse/controller/nats/HelloWorldController.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+@Component
+class HelloWorldController(
+    connection: Connection
+) {
+
+    init {
+        connection.createDispatcher { msg ->
+            val message = String(msg.data, StandardCharsets.UTF_8)
+            println("Received message: $message")
+            connection.publish("hi", msg.replyTo, "Hi $message".toByteArray())
```
Comment (ajax-sushkevych-s) [inline wm-system/src/main/kotlin/com/ajaxproject/warehouse/controller/nats/HelloWorldController.kt]: Where is Tony? :D

Comment (DmitryOnopriienko) [reply wm-system/src/main/kotlin/com/ajaxproject/warehouse/controller/nats/HelloWorldController.kt]: 😂😂😂

File: wm-system/src/main/kotlin/com/ajaxproject/warehouse/controller/nats/product/CreateProductNatsController.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                .createProduct(request.mapToDto()) // TODO validation (added on service layer, think about better impl)
+                .mapToProto()
+        )
+    }.getOrElse { exception ->
+        buildFailureResponse(exception)
+    }
+
+    fun buildSuccessResponse(product: Product): CreateProductResponse =
+        CreateProductResponse.newBuilder().apply {
+            successBuilder.setProduct(product)
```
Comment (ajax-shulepa-a) [inline wm-system/src/main/kotlin/com/ajaxproject/warehouse/controller/nats/product/CreateProductNatsController.kt]:
```suggestion
            successBuilder.product = product 
```
[nit]

File: wm-system/src/main/kotlin/com/ajaxproject/warehouse/dto/ProductDataDto.kt (hunk 4 above)
Comment (ajax-shulepa-a) [reply wm-system/src/main/kotlin/com/ajaxproject/warehouse/dto/ProductDataDto.kt]: to make it's more beautiful you can replace apply with also and there will be it.id = id. but up to you

**Review by ajax-shulepa-a — APPROVED** [review]

**Pull request merged**